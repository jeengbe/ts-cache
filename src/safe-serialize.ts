export interface SafeJsonOptions {
  /**
   * Specifies whether `undefined` values are allowed for object properties during serialization.
   * - If `true`, properties with `undefined` values will be serialized.
   * - If `false`, if any object contains a property with an `undefined` value, the function will throw an error.
   *
   * In most cases, `{ }` and `{ key: undefined }` are considered equivalent. This option allows for
   * greater control in situations they are not.
   *
   * @default true
   */
  allowUndefined?: boolean;

  /**
   * An array of types that will be whitelisted during serialization. These types will be treated
   * as serializable, even if they are normally excluded by default (e.g., `Date`, `Promise`, `Map`, etc.).
   * - You can specify built-in types like `Date`, `Promise`, or any custom class that you want to be serialized.
   *
   * This allows for custom handling of specific types that would otherwise require special serialization logic.
   * Checks are performed using the `instanceof` operator.
   *
   * @default []
   */
  // eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
  whitelistTypes?: Function[];
}

/**
 * Serializes a value to JSON with strict structural integrity.
 *
 * This function converts a given value to a JSON string while ensuring that the resulting
 * string, when parsed back with `JSON.parse`, produces a value that is structurally identical
 * to the original. If any part of the value would be transformed, omitted, or lose information
 * during the JSON conversion process (e.g. Dates, Promises, functions, cyclic references, etc.),
 * the function throws an error.
 */
export function safeJsonSerialize(
  value: unknown,
  options: SafeJsonOptions = {},
): string {
  if (!isFullyJsonSerializable(value, options)) {
    throw new Error('Value is not fully JSON-serializable.', {
      cause: {
        value,
      },
    });
  }

  return JSON.stringify(value);
}

function isFullyJsonSerializable(
  value: unknown,
  options: SafeJsonOptions,
  seen = new WeakSet<object>(),
): boolean {
  switch (typeof value) {
    case 'boolean':
    case 'string':
      return true;
    case 'number':
      return Number.isFinite(value);
    case 'bigint':
    case 'undefined': // if 'allowUndefined' is enabled, this case is checked in the 'object' branch directly
    case 'function':
    case 'symbol':
      return false;
    case 'object': {
      if (value === null) return true;

      if (seen.has(value)) return false;
      seen.add(value);

      if (Array.isArray(value)) {
        return value.every((item) =>
          isFullyJsonSerializable(item, options, seen),
        );
      }

      if (options.whitelistTypes?.some((type) => value instanceof type)) {
        return true;
      }

      // Check if it's a plain object: either has Object.prototype or null as its prototype.
      // Do this to prevent serializing Date, Promise, etc.
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const proto = Object.getPrototypeOf(value);
      if (proto !== Object.prototype && proto !== null) return false;

      return Object.keys(value).every((key) => {
        const keyValue = value[key as keyof typeof value] as unknown;

        if (keyValue === undefined) {
          return options.allowUndefined ?? true;
        }

        return isFullyJsonSerializable(keyValue, options, seen);
      });
    }
  }
}
