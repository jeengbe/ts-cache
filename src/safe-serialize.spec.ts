import { SafeJsonOptions, safeJsonSerialize } from './safe-serialize';

function makeCyclicObject() {
  const obj: any = {};
  obj.obj = obj;
  return obj;
}

class CustomType {
  constructor(public value: string) {}
}

describe('safeJsonSerialize', () => {
  allows('a boolean', true);
  allows('a string', 'hello');

  allows('a finite number', 42);
  disallows('an infinite number', Infinity);
  disallows('NaN', NaN);

  disallows('a BigInt', BigInt(42));
  disallows('undefined', undefined);
  disallows('a function', () => {});
  disallows('a symbol', Symbol('foo'));

  allows('null', null);
  allows('an empty object', {});
  allows('an object with string keys', { key: 'value' });
  allows('an object with number keys', { 42: 'value' });
  describe("an object with 'undefined' values", () => {
    allows('by default', { key: undefined });
    allows(
      'with allowUndefined: true',
      { key: undefined },
      { allowUndefined: true },
    );
    disallows(
      'with allowUndefined: false',
      { key: undefined },
      { allowUndefined: false },
    );
  });

  allows('an array', [1, 2, 3]);
  allows('an empty array', []);
  disallows('an array with undefined', [1, undefined, 3]);

  disallows('a date object', new Date());
  disallows('a promise object', Promise.resolve());
  allows('a plain object with null prototype', Object.create(null));
  allows('a date if whitelisted', new Date(), { whitelistTypes: [Date] });
  disallows("custom types that aren't whitelisted", new CustomType('foo'));
  allows('custom types that are whitelisted', new CustomType('foo'), {
    whitelistTypes: [CustomType],
  });
  disallows('an object with a cyclic reference', makeCyclicObject());

  function allows(
    description: string,
    value: unknown,
    options?: SafeJsonOptions,
  ) {
    it(`should serialize ${description}`, () => {
      expect(safeJsonSerialize(value, options)).toMatchSnapshot();
    });
  }

  function disallows(
    description: string,
    value: unknown,
    options?: SafeJsonOptions,
  ) {
    it(`should not serialize ${description}`, () => {
      expect(() => safeJsonSerialize(value, options)).toThrow(
        'Value is not fully JSON-serializable.',
      );
    });
  }
});
