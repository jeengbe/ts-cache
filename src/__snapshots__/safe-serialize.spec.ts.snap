// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`safeJsonSerialize an object with 'undefined' values should serialize by default 1`] = `"{}"`;

exports[`safeJsonSerialize an object with 'undefined' values should serialize with allowUndefined: true 1`] = `"{}"`;

exports[`safeJsonSerialize should serialize a boolean 1`] = `"true"`;

exports[`safeJsonSerialize should serialize a date if whitelisted 1`] = `""1970-01-01T00:00:00.123Z""`;

exports[`safeJsonSerialize should serialize a finite number 1`] = `"42"`;

exports[`safeJsonSerialize should serialize a plain object with null prototype 1`] = `"{}"`;

exports[`safeJsonSerialize should serialize a string 1`] = `""hello""`;

exports[`safeJsonSerialize should serialize an array 1`] = `"[1,2,3]"`;

exports[`safeJsonSerialize should serialize an empty array 1`] = `"[]"`;

exports[`safeJsonSerialize should serialize an empty object 1`] = `"{}"`;

exports[`safeJsonSerialize should serialize an object with number keys 1`] = `"{"42":"value"}"`;

exports[`safeJsonSerialize should serialize an object with string keys 1`] = `"{"key":"value"}"`;

exports[`safeJsonSerialize should serialize custom types that are whitelisted 1`] = `"{"value":"foo"}"`;

exports[`safeJsonSerialize should serialize null 1`] = `"null"`;
